File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Stores the last seen index of each character
        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0;   // start of window
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);

            // If character is already in window, move the left pointer
            if (map.containsKey(ch) && map.get(ch) >= left) {
                left = map.get(ch) + 1;
            }

            // Update the last seen index
            map.put(ch, right);

            // Calculate window size
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        int left = 0;

        unordered_map<char, int> map;

        char* arr = &s[0];

        int right = 0;

        for (; right < s.length(); right++) {

            int idx = map.count(arr[right]) ? map[arr[right]] : -1;

            if (idx != -1 && idx >= left) {

                res = max(res, right - 1 - left + 1);

                left = idx + 1;
            }

            map[arr[right]] = right;
        }

        return max(res, right - 1 - left + 1);
    }
};